Resumable Upload: Add guidance on upload creation strategy

Acconut opened the pull request:
The definition of upload creation requests allows the client to include the entire file, some file data or no data at all. While we provide this flexibility, we don't give guidance on when which approach is best suited.

This PR adds this missing piece and explains how an optimistic client can attempt to upload the file in one request and how a careful client uses a separate request for creating the upload. In addition, it also mentions how to transparently upgrade to resumable uploads, which wasn't explained in detail before.

Let me know if you have any feedback on this.

guoye-zhang commented:
Are "Optimistic Upload Creation" and "Upgrading To Resumable Uploads" the exact same way?

Maybe the 3rd way could be checking `Upload-Limit` first before uploading?

Acconut commented:
> Are "Optimistic Upload Creation" and "Upgrading To Resumable Uploads" the exact same way?

The way they are described right now they are similar, yes. In theory, an optimistic upload creation can also be used where the client doesn't upload the entire file, but just a large part of it (think 500 MB of a 10 GB file). When upgrading a non-resumable request to a resumable upload, this act of division is not allowed. But I am not sure if we should go into that much detail here. If not, then we can merge these two sections.

> Maybe the 3rd way could be checking `Upload-Limit` first before uploading?

Upload limits are a good point! With the careful upload, the client also knows the limits before sending the first PATCH request. The optimistic client must be prepared to receive a 104 response with limits that it has to adapt to.

We can mention that a careful client may also send an OPTIONS request before hand if it deems that necessary.

smatsson commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -512,6 +512,34 @@ If the client knows the integrity digest of the content from an upload creation
+
+## Optimistic Upload Creation
+
+An "optimistic upload creation" can be used if the client knows or assumes with a high degree of confidence that the server supports resumable uploads under the target URI and that the client is capable of handling and processing interim responses. The upload creation request then includes the full file because the client assumes that the file will be transferred without interruptions.
```
>... because the client assumes that the file will be transferred without interruptions.

> ... because the client assumes that the file will be transferred without interruptions or resumed if an interruption occurs.

smatsson reviewed (commented):
Just a minor nit. Otherwise I think this is a very good addition and agree with @guoye-zhang   that we should add some text with regards to the `Upload-Limit`.

Proposal for optimistic: The client needs to adapt to the `Upload-Limit` header provided in the `104 (Upload Resumption Supported)` intermediate response and fail the request if the file is larger than the maximum allowed size.

Proposal for Careful: The client can use the  `Upload-Limit` header provided in the `104 (Upload Resumption Supported)` to verify that the file is allowed to be uploaded before sending any data.

guoye-zhang commented:
I think upload-limit can just be enforced on the server side. If it sends a 104 and a 5xx response, the client would automatically retry. We have deployed a client-side implementation that is not aware of the upload-limit header field.

martinthomson commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -512,6 +512,34 @@ If the client knows the integrity digest of the content from an upload creation
 
+# Upload Strategies
+
+The definition of the upload creation request ({{upload-creation}}) provides the client with flexibility to choose whether the file is fully or partially transferred in the first request, or if no file data is included at all. The decision which behavior is best largely depends on the client's capabilities, its intention to avoid data transmission, and its knowledge about the server's support for ...
```
```suggestion
The definition of the upload creation request ({{upload-creation}}) provides the client with flexibility to choose whether the file is fully or partially transferred in the first request, or if no file data is included at all. Which behavior is best largely depends on the client's capabilities, its intention to avoid data transmission, and its knowledge about the server's support for resumable uploads.
```

martinthomson commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -512,6 +512,34 @@ If the client knows the integrity digest of the content from an upload creation
+
+An "optimistic upload creation" can be used if the client knows or assumes with a high degree of confidence that the server supports resumable uploads under the target URI and that the client is capable of handling and processing interim responses. The upload creation request then includes the full file because the client assumes that the file will be transferred without interruptions.
+
+The benefit of this method is that if the upload creation request succeeds, the file was transferred in a single request avoiding additional round trips. A drawback is that the client may be restricted in its ability to resume that upload. If the upload was interrupted before the client received a `104 (Upload Resumption Supported)` intermediate response with the upload URL, the client cannot res ...
```
```suggestion
The benefit of this method is that if the upload creation request succeeds, the file was transferred in a single request without additional round trips.

A possible drawback is that the client might be unable to resume an upload. If an upload is interrupted before the client received a `104 (Upload Resumption Supported)` intermediate response with the upload URL, the client cannot resume that upload due to the missing upload URL. The intermediate response might not be received if the interruption happens too early in the message exchange, the server does not support sending the 104 (Upload Resumption Supported) intermediate response, or an intermediary dropped the intermediate response. Without a 104 response, the client needs to either treat the upload as failed or retry the entire upload creation request if this is allowed by the application.
```

I'll note that if the 104 doesn't have time to arrive, then the client isn't really losing much progress.  As long as they are able to start over, this seems like a very reasonable approach.

martinthomson commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -512,6 +512,34 @@ If the client knows the integrity digest of the content from an upload creation
+
+The benefit of this method is that if the upload creation request succeeds, the file was transferred in a single request avoiding additional round trips. A drawback is that the client may be restricted in its ability to resume that upload. If the upload was interrupted before the client received a `104 (Upload Resumption Supported)` intermediate response with the upload URL, the client cannot res ...
+
+The "optimistic upload creation" attempts to upload the file in one request at the potential expense of either not finishing the upload at all or re-transmitting file data again.
+
```
```suggestion
```

This is redundant.

martinthomson commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -512,6 +512,34 @@ If the client knows the integrity digest of the content from an upload creation
+
+The approach specified in this document allows clients and servers to automatically upgrade non-resumable uploads to resumable ones. In a non-resumable upload, the file is transferred in a single request, usually `POST` or `PUT`, without any ability to resume from interruptions. The client can offer the server to upgrade such an upload to a resumable upload. If supported by the server, the file t ...
+
+To perform an upgrade, the client adds the `Upload-Complete: ?1` header field to the original request, which attempts to transmit the entire file in one request. A server that supports resumable uploads at the target URI, can create a resumable upload resource and send its upload URL in a `104 (Upload Resumption Supported)` intermediate response for the client to resume the upload after interrupt ...
```
```suggestion
To perform an upgrade, the client adds the `Upload-Complete: ?1` header field to the original request, which attempts to transmit the entire file in one request. A server that supports resumable uploads at the target URI, can create a resumable upload resource and send its upload URL in a 104 (Upload Resumption Supported) intermediate response for the client to resume the upload after interruptions.

A server that does not support resumable uploads or does not want to upgrade to a resumable upload for this request can ignore the `Upload-Complete: ?1` header and treat the request as a non-resumable upload.
```

martinthomson commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -512,6 +512,34 @@ If the client knows the integrity digest of the content from an upload creation
+
+This approach best suited if the client cannot receive intermediate responses, e.g. due to a limitation in the provided HTTP interface, or if large files are transferred where the cost of the additional request is miniscule compared to the effort of transferring the large file itself.
+
+## Upgrading To Resumable Uploads
```
This seems indistinguishable from the optimistic approach to me.  The use of Upload-Complete might distinguish it, but I don't see how these are functionally different.

Acconut replied:
You're correct. I merged them.

martinthomson reviewed (commented):
Some editorial suggestions, but one big question:

What is the material difference between the first and third strategy?  Could these be the same strategy?

Acconut commented:
Thanks for all of your comments! I updated the PR to put the section about upgrading to resumable uploads under the optimistic upload creation since they are similar. I also added a brief mention of Upload-Limit and shortened the text a bit.

Let me know if you think the text needs more adjustments.

guoye-zhang commented on draft-ietf-httpbis-resumable-upload.md:
```diff
@@ -512,6 +512,36 @@ If the client knows the integrity digest of the content from an upload creation
+
+## Optimistic Upload Creation
+
+An "optimistic upload creation" can be used if the client assumes that the server supports resumable uploads under the target URI and that the client is capable of handling and processing interim responses. The upload creation request then includes the full file because the client anticipates that the file will be transferred without interruptions or resumed if an interruption occurs.
```
> assumes that the server supports resumable uploads under the target URI

This seems to be a bit in conflict with the "Upgrading To Resumable Uploads" subsection now that they are merged. Maybe we can say "assumes that the server supports resumable uploads under the target URI, or has no knowledge of the server support", but then that sentence doesn't really say anything.

"Careful Upload Creation" is where the server must support resumable uploads. "Optimistic Upload Creation" is OK if the server doesn't.

Acconut replied:
Yes, the client's knowledge about the server support was a difference of the two sections at first. Now that they are merged, the text gets a bit more unwieldy, but a added https://github.com/httpwg/http-extensions/commit/de5992146b5957168dc2c7e0af2a2dbe64086ab5 to correct the statements about the client's knowledge. What do you think? Is that sufficient or do we need more significant changes?

guoye-zhang replied:
Looks good, thanks
